### PR TITLE
[FLINK-22944][state] Optimize writing state changelog

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

This PR optimizes writing of state changes in `ChangelogKeyedStateBackend`.
Currently, state name and type are written with every change.
With this change, only a short ID of state is written every time; the mapping from this ID to name and type is written before the first write.

The change is split into three commits to ease reviewing. They could be squashed together upon merging.

An alternative approach would be to sort changes in each write batch and write common part once.
That would minimize writing of keys and namespaces.
However, it has some drawbacks:
 - writing again in every batch
 - more complex writer interface - it must be aware of ns/keys as well as serializers (they differ per call because different state of of the same backend have different serializers)
 - more complex format and read/write logic
 - harder to re-use writer for non-backend purposes
 - harder to re-use sorting code in other writer implementations

Because of the above drawbacks the mapping approach was chosen.

## Verifying this change

Correctness is covered by the existing changelog backend tests.
Optimization is not covered.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

File: flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java [hunk 1]
```diff
@@ -506,7 +517,9 @@ public void notifyCheckpointAborted(long checkpointId) throws Exception {
                         stateChangelogWriter,
                         meta,
                         stateDesc.getTtlConfig(),
-                        stateDesc.getDefaultValue());
+                        stateDesc.getDefaultValue(),
+                        ++lastCreatedStateId);
```
Comment (Myasuka) [inline flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:521]: I wonder how the `lastCreatedStateId` work with restored state? Will it increase to the restored state id?

**Review by Myasuka — CHANGES_REQUESTED** [review]
My main concern is how to handle the rescaling case? As same state could have different short state id in different subtasks. Since we use key-group as the core design in rescaling, maybe we could combine it with group info.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendHandle.java [hunk 2]
```diff
@@ -48,20 +49,25 @@
 
     List<ChangelogStateHandle> getNonMaterializedStateHandles();
 
+    UUID getBackendID();
```
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendHandle.java:52]: What the purpose of UUID acts here?

File: flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java (hunk 1 above)
Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:521]: There is no connection between "old" state id (before recovery) and the "new" one (after recovery).
Recovery doesn't depend on the new state id as the mapping (from id to name) is written in snapshot (though recovery might increment `lastCreatedStateId`).
New snapshot after recovery will not depend on the old id because the mapping will be written at least once after recovery.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendHandle.java (hunk 2 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendHandle.java:52]: UUID identifies the backend, which allows to avoid collisions of state id across different backends on recovery (e.g. in case of downscaling).

**Comment by rkhachatryan** [comment]
Thanks for the review @Myasuka 
> My main concern is how to handle the rescaling case? As same state could have different short state id in different subtasks. 

Good point. I addressed this with the backend UUID (discussed in [this thread](https://github.com/apache/flink/pull/17203#discussion_r707877114)).

> Since we use key-group as the core design in rescaling, maybe we could combine it with group info.

I think using key groups could be more fragile than backend UUID because they can be re-assigned multiple times (with multiple rescalings). Besides that, that would require more space in snapshot because state name would be written the number of times the number of key group assigned.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendHandle.java (hunk 2 above)
Comment (Myasuka) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateBackendHandle.java:52]: Thanks for your explanation. I think we could add a basic java doc on it to describe this.

**Review by Myasuka — COMMENTED** [review]
Thanks for the reply. From my understanding, it seems lacks of a unit test to verify the rescaling case in which we introduce the backend uuid in changelog state handle. I think this logic deserve a unit test.

**Comment by rkhachatryan** [comment]
Regarding the tests, there are already in `ChangelogDelegateHashMapTest` and similar, e.g. `testKeyGroupSnapshotRestoreScaleDown`, `testKeyGroupSnapshotRestoreScaleUp`, `testKeyGroupsSnapshotRestoreScaleUpUnEvenDistribute`, etc.
which seems enough, WDYT?

File: flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java [hunk 3]
```diff
@@ -190,6 +194,19 @@
 
     private final ExecutorService asyncOperationsThreadPool;
 
+    /**
+     * Identifies this backend. On restore, when snapshots of multiple backends are present
+     * (downscaling case) this allows to differentiate between state mappings of different backends.
```
Comment (Myasuka) [inline flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:199]: Actually, upscaling case could also cause this case (such as change parallelism 2 to 3).

File: flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogBackendRestoreOperation.java [hunk 4]
```diff
@@ -80,14 +89,19 @@
...
             StateChangelogHandleReader<T> changelogHandleReader,
-            ClassLoader classLoader)
+            ClassLoader classLoader,
+            Map<UUID, Map<Short, StateID>> metadataByBackend)
             throws Exception {
         for (ChangelogStateHandle changelogHandle :
                 backendHandle.getNonMaterializedStateHandles()) {
+            Map<Short, StateID> stateIds =
+                    metadataByBackend.computeIfAbsent(
+                            backendHandle.getBackendID(), ign -> new HashMap<>());
```
Comment (Myasuka) [inline flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogBackendRestoreOperation.java:99]: To answer the question that whether to add a unit test, I checked all tests under `flink-statatebackend-changelog` module, and found that the `stateIds` here is always empty. This could be explained that we would not restore two state handles from previous same statebackend, and the statd-ids are updated during `applyOperation`.
If so, do we still need the API of `getBackendUID`?

Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogBackendRestoreOperation.java]: You are right, I've removed backend ID (both from handle and from backend itself).

I started to [refactor](https://github.com/rkhachatryan/flink/tree/refactor-backend-test) rescaling tests to add more snapshots for the same backend and then shuffle them (so that metadata from one backend could overwrite metadata of another).

But I realized that it won't break recovery, because each snapshot (i.e. `ChangelogStateBackendHandle`) contains metadata in the beginning; and reading of each such snapshot is "atomic" (i.e. doesn't interlieave with reading other snapshots).

File: flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogBackendRestoreOperation.java [hunk 5]
```diff
@@ -89,14 +82,11 @@
             ChangelogKeyedStateBackend<?> backend,
             ChangelogStateBackendHandle backendHandle,
             StateChangelogHandleReader<T> changelogHandleReader,
-            ClassLoader classLoader,
-            Map<UUID, Map<Short, StateID>> metadataByBackend)
+            ClassLoader classLoader)
             throws Exception {
+        Map<Short, StateID> stateIds = new HashMap<>();
```
Comment (Myasuka) [inline flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogBackendRestoreOperation.java:87]: The equivalent change should be creating the stateIds within the for loop, otherwise, the stateIds would contain short state id created by other state backends. Though we would update the state id during meta restore, I'm afraid that this might not be safe if we change implementation of updating the meta in the future.

Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogBackendRestoreOperation.java:87]: As discussed offline:
The loop iterates over the log segments of a single backend (`backendHandle.getNonMaterializedStateHandles()`)
So it would be incorrect to re-create the map inside the loop, because earlier log segments might contain metadata needed for the later ones.
And there should be no data from other backends during the whole loop, because `backendHandle` comes from a single backend.

**Review by Myasuka — APPROVED** [review]
LGTM. 
It seems the CI failed due to https://issues.apache.org/jira/browse/FLINK-24305

**Comment by rkhachatryan** [comment]
Thanks for the through review @Myasuka!

I've squashed the commits and will merge the PR after the build succeeds.
